velox/common/memory/MemoryArbitration.h

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/base/SuccinctPrinter.h"
+
+namespace facebook::velox::memory {
+
+class MemoryPool;
+
+/// The memory arbitrator interface. There is one memory arbitrator object per
+/// memory manager which is responsible for arbitrating memory usage between the
+/// root memory pools allocated from the memory manager. Different memory
+/// arbitrator implementation enforce different arbitration policies. For
+/// example, Prestissimo needs a fixed size memory arbitration policy to enforce
+/// the memory isolation between running queries on the same worker node. Each
+/// query throws memory exceeding exception when exceeding the fixed query
+/// memory limit. Sapphire needs a shared memory arbitration policy to achieve
+/// dynamic memory sharing across queries so that one query can use more than
+/// its fair share during the time that the system has free available memory and
+/// the excess memory can be reclaimed back on-demand later through memory
+/// reclaim mechanism such as spilling techniques.
+class MemoryArbitrator {
+ public:
+  /// Defines the kind of memory arbitrators.
+  enum class Kind {
+    /// Used by Prestissimo to enforce fixed size memory isolation across
+    /// queries running on the same node.
+    kFixed,
+    /// Used by Sapphire to achieve dynamic memory sharding across quries
```

Comment:
sharding -> sharing?

quries -> queries

> running on the same node.

My impression is that Presto-on-Spark runs each query in a separate process and doesn't run multiple queries within the same process. Is this not the case?

Reply:
Yes, it is supposed to be but this can be generic enough to apply for the use case with multiple running queries. If there is only one query, then we always reclaim from the query itself.